sys/kern/cred_checks.c

```diff
@@ -0,0 +1,32 @@
+#include <sys/cred.h>
+#include <sys/proc.h>
+#include <sys/errno.h>
+#include <sys/mutex.h>
+
+int cred_cansignal(proc_t *target, cred_t *cred) {
+  assert(mtx_owned(&target->p_lock));
+
+  if (cred->cr_euid == 0)
+    return 0;
+
+  if (cred->cr_ruid != target->p_cred.cr_ruid &&
+      cred->cr_ruid != target->p_cred.cr_suid &&
+      cred->cr_euid != target->p_cred.cr_ruid &&
+      cred->cr_euid != target->p_cred.cr_suid)
+    return EPERM;
+
+  return 0;
+}
+
+int proc_cansignal(cred_t *cred, session_t *s, proc_t *target, signo_t sig) {
+  assert(mtx_owned(&target->p_lock));
```

Comment:
I would like the signature of this function to look like this:
```c
/* Checks whether the current process has permission to send `sig` to `target` process. */
int proc_cansignal(proc_t *target, signo_t sig)
```
We can require `all_proc_mtx` to be held in addition to `&target->p_lock` -- then we will be able to safely access the current process's session.

Reply:
Applied your suggestion.